rfcs/20200612-stream-executor-c-api.md

```diff
@@ -0,0 +1,502 @@
+            we will try to keep them stable (in a best-effort manner) until the new 
+            TensorFlow stack is available.
+
+# Design Proposal
+
+## StreamExecutorInterface
+
+[StreamExecutorInterface](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=166?q=StreamExecutorinterface)
+is quite large and some of its methods are only sporadically used. Therefore, we
+plan to wrap only a subset of key StreamExecutorInterface functionality.
+
+See proposed C API below:
+
+```cpp
+#include <stddef.h>
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef SE_Stream_st* SE_Stream;
+typedef SE_Event_st* SE_Event;
+typedef SE_Timer_st* SE_Timer;
+typedef TF_Status* (*TF_StatusCallbackFn)(void*);
+
+#ifndef TF_BOOL_DEFINED
+#define TF_BOOL unsigned char
+#endif // TF_BOOL_DEFINED
+
+#ifndef TF_OFFSET_OF_END
+#define TF_OFFSET_OF_END(TYPE, MEMBER) (offsetof(TYPE, MEMBER) + sizeof(((TYPE *)0)->MEMBER))
+#endif // TF_OFFSET_OF_END
+
+typedef struct SE_PlatformId {
+ size_t struct_size;
+ void* id;  // aka stream_executor::Platform::Id
+} SE_PlatformId;
+
+#define SE_PLATFORMID_STRUCT_SIZE TF_OFFSET_OF_END(SE_PlatformId, id)
+
+typedef struct SE_TimerFns {
+ size_t struct_size;
+ uint64_t (*nanoseconds)(SE_Timer timer);
+ uint64_t (*microseconds)(SE_Timer timer);
+} SE_TimerFns;
+
+#define SE_TIMER_STRUCT_SIZE TF_OFFSET_OF_END(SE_Timer, microseconds)
+
+typedef struct SE_AllocatorStats {
+  size_t struct_size;
+  int64_t num_allocs;
+  int64_t bytes_in_use;
+  int64_t peak_bytes_in_use;
+  int64_t largest_alloc_size;
+
+  int8_t has_bytes_limit;
+  int64_t bytes_limit;
+
+  int64_t bytes_reserved;
+  int64_t peak_bytes_reserved;
+
+  int8_t has_bytes_reservable_limit;
+  int64_t bytes_reservable_limit;
+
+  int64_t largest_free_block_bytes;
+} SE_AllocatorStats;
+
+#define SE_ALLOCATORSTATS_STRUCT_SIZE TF_OFFSET_OF_END(SE_AllocatorStats, largest_free_block_bytes)
+
+typedef enum SE_EventStatus {
+  SE_EVENT_UNKNOWN,
+  SE_EVENT_ERROR,
+  SE_EVENT_PENDING,
+  SE_EVENT_COMPLETE,
+} SE_EventStatus;
+
+typedef struct SE_Options {
+  size_t struct_size;
+  int32_t ordinal;
+};
+
+#define SE_OPTIONS_STRUCT_SIZE TF_OFFSET_OF_END(SE_Options, ordinal)
+
+typedef struct SE_Device {
+  size_t struct_size;
+  char* name;
+  size_t name_len;
+
+  // Device vendor can store handle to their device representation
+  // here.
+  void* device_handle;
+
+  // Any kind of data that plugin device might want to store.
+  void* data;
+};
+
+#define SE_DEVICE_STRUCT_SIZE TF_OFFSET_OF_END(SE_Device, data)
+
+typedef struct SE_StreamExecutor {
+  size_t struct_size;
+
+  /*** ALLOCATION CALLBACKS ***/
+  // Synchronously allocates size bytes on the underlying platform and returns
+  // a DeviceMemoryBase representing that allocation. In the case of failure,
+  // nullptr is returned.
+  TF_DeviceMemoryBase* (*allocate)(
+      SE_Device* se, uint64_t size, int64_t memory_space);
```

Comment:
memory_space needs to be an enum or a comment on domain of values accepted. It seems to originate from [enum MemorySpace](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/xla/service/memory_space_assignment.h;drc=fc49cbb2adf0b69f843f7bf904978648bffbe268;bpv=1;bpt=1;l=349?q=memory_space_assignment.h&gsn=MemorySpace&gs=kythe%3A%2F%2Fgithub.com%2Ftensorflow%2Ftensorflow%3Flang%3Dc%252B%252B%3Fpath%3Dtensorflow%2Fcompiler%2Fxla%2Fservice%2Fmemory_space_assignment.h%236xcM2GFLtrjH9IrmPLzqjHHeEL8C75jUFl84E30oVvA).

Reply:
I see two `MemorySpace` structs in TF codebase: https://osscs.corp.google.com/search?q=%22MemorySpace%20%7B%22

I am going to ask around and see what it stands for.
Right now, I am guessing `memory_space` would be whatever `PluggableDevice` passes it in during allocation. At the same time `PluggableDevice` might not use it at all.

Reply:
Seems like there is no fixed struct to use for memory_space yet.
So,  I added a comment indicating it should be `0` for now and reserved for potential future usecase.